File: algorithms/Yandex/algorithms_training_4/sorts/Python/mergeSort.py
def merge(res_arr:list, left_arr: list, right_arr: list) -> list:
    index = left_index = right_index = 0
    while left_index < len(left_arr) and right_index < len(right_arr):
        if left_arr[left_index] <= right_arr[right_index]:
            res_arr[index] = left_arr[left_index]
            left_index += 1
        else:
            res_arr[index] = right_arr[right_index]
            right_index += 1
        index += 1
    for i in range(left_index, len(left_arr)):
        res_arr[index] = left_arr[i]
        index += 1
    for i in range(right_index, len(right_arr)):
        res_arr[index] = right_arr[i]
        index += 1
    return res_arr


def mergeSort(arr: list) -> list:
    if len(arr) <= 1: return arr
    return merge(arr, mergeSort(arr[:len(arr) // 2]), mergeSort(arr[len(arr) // 2:]))

input()
user_arr = list(map(int, input().split()))
print(*mergeSort(user_arr))

File: algorithms/Yandex/algorithms_training_4/sorts/С++/mergeSort.cpp
#include <iostream>
#include <vector>

using std::cin, std::cout;

void merge(std::vector<int>& res, std::vector<int>& left, std::vector<int>& right) {
    int index = 0, left_index = 0, right_index = 0;
    while (left_index < left.size() && right_index < right.size()) {
        if (left[left_index] <= right[right_index]) res[index++] = left[left_index++];
        else res[index++] = right[right_index++];
    }
    for (int i = left_index; i != left.size(); ++i) res[index++] = left[i];
    for (int i = right_index; i != right.size(); ++i) res[index++] = right[i];
}


std::vector<int> mergeSort(std::vector<int>& vec) {
    if (vec.size() <= 1) return vec;
    std::vector<int> left {vec.begin(), vec.begin() + vec.size() / 2};
    std::vector<int> right {vec.begin() + vec.size() / 2, vec.end()};
    mergeSort(left);
    mergeSort(right);
    merge(vec, left, right);
    return vec;
}


int main() {
    int size;
    cin >> size;

    if (size == 0) {
        cout << '\n';
        return 0;
    }
    std::vector<int> vec(size);
    for (int& elem: vec) cin >> elem;
    mergeSort(vec);
    for (int i = 0; i != vec.size() - 1; ++i) cout << vec[i] << ' ';
    cout << vec[vec.size() - 1];
    return 0;
}

